### ISSUE TITLE
[Request] Add the option to "Maximize Plan" for the current skill

### ISSUE THREAD

**Issue by Tharnadar** [pyfa-org/Pyfa#2132]
Greetings, i would like to have the option to export the skill plan for all the involved skills at level V for the current fit.
I tried to ask this functionality to EveMon, but they rejected the request because they do not have a full Dogma engine (https://github.com/peterhaneve/evemon/issues/198).
Thank you very much.

**Comment by DarkFenX** [comment]
This is not very easy question to be fair.

1) What about fitting skills? For example, your fit fits with CPU management 3, does it have to be exported as 3 or 5? If we optimize for "fits just enough", solution might be not that easy with multiple skills affecting fitting (CPU management, weapon upgrades, power grid upgrades, electronics upgrades, launcher rigging, drones rigging all affect CPU for example).
2) What about skills which provide no apparent benefit? For example, tactical shield manipulation 5 in some circumstances can be considered harmful.
3) What about skills which are rarely needed? For example, almost any recon/logistics and other ships with signal amplifiers or auto targeting systems benefit from advanced target management 5. Should it also be included?

I think we can just embrace one way or another for 2) and 3), but optimized solutions for 1) are very expensive to calculate with current pyfa engine

**Comment by Tharnadar** [comment]
Thank you DarkFenX for the fast reply, and I know it's not an easy implementation, and please be patient with my english that's not so good.
I try to describe what I do when I try to maximize a skillplan for a given fit: I cycle every installed modules, right click, "Change Module Skills", take note of skill name, then I go to Evemon and I add it to plan at level V; after that i do the same for charges; and at the end for the ship/hull.

**Comment by DarkFenX** [comment]
This way it would be easy enough, but not very useful in many cases. I will look into it, but later.

**Comment by Tharnadar** [comment]
Hi DarkFenX, first time programming in python, i've made a small utiliy to achieve the solution, it's basically the copy of Character.checkRequirements with few modifies, I hope it helps.

```
import itertools

from functools import reduce
from eos.saveddata.damagePattern import DamagePattern
from eos.utils.stats import RRTypes, DmgTypes
from gui.utils.numberFormatter import formatAmount

from eos.saveddata.implant import Implant as es_Implant
from eos.saveddata.character import Character as es_Character, Skill
from eos.saveddata.module import Module as es_Module
from eos.const import FittingSlot as es_Slot
from eos.saveddata.fighter import Fighter as es_Fighter

def exportMaxSkillPlan(fit, callback):
    # toCheck = []
    reqs = {}
    skills = []
    for thing in itertools.chain(fit.modules, fit.drones, fit.fighters, (fit.ship,), fit.appliedImplants, fit.boosters):
        if isinstance(thing, es_Module) and thing.slot == es_Slot.RIG:
            continue
        for attr in ("item", "charge"):
            if attr == "charge" and isinstance(thing, es_Fighter):
                # Fighter Bombers are automatically charged with micro bombs.
                # These have skill requirements attached, but aren't used in EVE.
                continue
            subThing = getattr(thing, attr, None)
            subReqs = {}
            if subThing is not None:
                if isinstance(thing, es_Fighter) and attr == "charge":
                    continue
                _checkRequirements(fit.character, subThing, subReqs, skills)
                if subReqs:
                    reqs[subThing] = subReqs
    #return reqs
    levels = ["I", "II", "III", "IV", "V"]
    text = "\n"
    for skill in skills:
        index = 0
        while index < 5:
            text += skill + " " + levels[index] + "\n"
            index += 1

    if callback:
        callback(text)
    else:
        return text


def _checkRequirements(char, subThing, reqs, skills):
    for req, level in subThing.requiredSkills.items():
        name = req.name
        ID = req.ID
        info = reqs.get(name)
        currLevel, subs = info if info is not None else 0, {}
        #if level > currLevel and (char is None or char.getSkill(req).level < level):
        reqs[name] = (level, ID, subs)
        _checkRequirements(char, req, subs, skills)
        if name not in skills:
            skills.append(name)
    return reqs
```

**Comment by DarkFenX** [comment]
This way you get maximized skill requirements and nothing else. Some skills are not direct requirements for anything but they help your ship - for example, Shield Management 5 (iirc it's not a requirement for any module).

**Comment by Tharnadar** [comment]
yes i know, but that's exaclty what i needed, the skillplan to maximize the skills for a given fit+ship... i guess the ship max skill is a subset of magic 14.
anyway i understand if you don't want to add this functionality to "official" pyfa; if i need the function i'll launch my pyfa version ;)

**Comment by DarkFenX** [comment]
Not really... i just do not want to do it right away, and i dislike half-assed features. I do not mean to offend you, but your implementation just would not work for your purpose, just because it does not include many skills which benefit a fit (like rapid firing 5 for fits with guns).
